tensorflow/compiler/jit/cluster_scoping_pass.cc

```diff
@@ -0,0 +1,148 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/compiler/jit/cluster_scoping_pass.h"
+
+#include "absl/algorithm/container.h"
+#include "absl/container/flat_hash_set.h"
+#include "absl/strings/str_cat.h"
+
+#include "tensorflow/compiler/jit/defs.h"
+#include "tensorflow/compiler/jit/xla_cluster_util.h"
+#include "tensorflow/core/framework/node_def_util.h"
+#include "tensorflow/core/graph/algorithm.h"
+
+namespace tensorflow {
+namespace {
+
+class ClusterScopingPassImpl {
+ public:
+  ClusterScopingPassImpl(Graph* graph,
+                         OptimizerOptions::GlobalJitLevel global_jit_level)
+      : graph_(graph),
+        global_jit_level_(global_jit_level),
+        unique_scope_id_(0) {}
+
+  Status Run();
+
+ private:
+  Status ScopingForPipelineStages();
+
+  size_t GetUniqueScopeId() { return unique_scope_id_++; }
+
+  void AddScopeToAllTransitivePredecessors(Node* start);
+
+  void AddScopeToAllTransitiveSuccessors(Node* start);
+
+ private:
+  Graph* graph_;
+  OptimizerOptions::GlobalJitLevel global_jit_level_;
+  size_t unique_scope_id_;
+};
+
+absl::optional<string> GetXlaInternalScope(Node* node) {
+  string scope;
+  if (GetNodeAttr(node->attrs(), kXlaInternalScopeAttr, &scope).ok()) {
+    return scope;
+  }
+
+  return absl::nullopt;
+}
+
+void SetXlaInternalScope(Node* node, StringPiece scope) {
+  node->AddAttr(kXlaInternalScopeAttr, scope);
+}
+
+// NB! We append a new scope as suffix to the _XlaInternalScope attribute
+// instead of overriding the old value.  In this way, we respect the original
+// scopes.  In other words, appending X to Y creates the conjunction of the
```

Comment:
I see where you are from. It is indeed confusing what it means to respect the "original" scope. And yes, getting both scopes is a good description.  I rewrote the comments by adopting your view and also giving a better explanation to my view. My view is that a unique scope in effect disallows merging a node to nodes in other scopes. So, scope conjunction preserves this semantic of disallowing merging.

In the below example, I'd like to point out that putting X and Y into different clusters is usually a more desirable result. Note that the long latency op is usually the `IteratorGetNext`, which processes input image with a lambda function. So, we do not want X to depend on IteratorGetNext.

`Unstage` -> `X` -> `Stage`;  `IteratorGetNext` -> `Y` -> `Stage`

On the other hand, `Stage`/`Unstage` themselves do not take a time too long to enqueue/dequeue. If you confuse them with long latency ops, it may be my fault as I (mistakenly) indicated so in a previous email. In contrast, Stage/Unstage is better treated as having semantics of wanted parallelism (instead of long latency.)